ql/cashflows/capflooredinflationcoupon.cpp

```diff
@@ -70,7 +70,7 @@ namespace QuantLib {
                          underlying->referencePeriodStart(),
                          underlying->referencePeriodEnd()),
       underlying_(underlying), isFloored_(false), isCapped_(false) {
-
+        // NOLINTNEXTLINE(clang-analyzer-optin.cplusplus.VirtualCall)
         setCommon(cap, floor);
```

Comment:
Instead of using the clang macro, does writing this as
```
        CappedFlooredYoYInflationCoupon ::setCommon(cap, floor);
```
silence the warning?  This way at least we'd be writing the current behavior explicitly.  In most cases I think that's what we wanted.

Reply:
Yes, that does make the warning go away, but it also effectively devirtualizes the call to `setCommon`, which might result in different behavior for anyone who has overridden it in a subclass (there aren't any subclasses in QuantLib itself). If we're going to do that then I think it's better to simply make `setCommon` non-virtual.

Reply:
Right, I wasn't clear — I'm only talking about constructors, which are the majority of cases and in which the call is already devirtualized.

As you say, this doesn't work for indirect calls, i.e., calls in methods that also happen to be called by constructors. 
 They will have to be managed differently.

Reply:
I'm not quite sure I follow. The [clang-analyzer-optin.cplusplus.VirtualCall](https://releases.llvm.org/14.0.0/tools/clang/tools/extra/docs/clang-tidy/checks/clang-analyzer-optin.cplusplus.VirtualCall.html) checks for any virtual functions called in a constructor or destructor, because this can result in unexpected behavior and is [generally discouraged](https://isocpp.org/wiki/faq/strange-inheritance#calling-virtuals-from-ctors).

In this case, the virtual method `setCommon` is called (directly, not indirectly) from the constructor of `CappedFlooredYoYInflationCoupon` so it's a clear violation of the guideline. The options are 1) suppress the warning, 2) devirtualize by specifying class prefix, 3) devirtualize by removing virtual function specifier.

Reply:
I meant that when virtual methods are called in a constructor, devirtualizing them by specifying the class prefix doesn't change the existing behavior: they are devirtualized already, which is the whole reason the behavior can be unexpected.  Specifying the class prefix only makes it explicit.

Reply:
Yes I think we are in agreement. So the question is, do you think it's better to add the class prefix wherever possible, or to try and remove the virtual function specifier even if it slightly changes the public interface? I suppose we can keep the suppression where neither of those two options are possible.

Reply:
I'd go for the class prefix.